src/python/pants/init/specs_calculator.py

```diff
@@ -125,6 +143,36 @@ def create(
                 filesystem_specs.append(FilesystemLiteralSpec(file_name))
             else:
                 address_specs.append(SingleAddress(address.spec_path, address.target_name))
+
+
+        if query_expr_strings:
+            # TODO(#7346): deprecate --owner-of and --changed-* in favor of --query versions, allow
+            # pipelining of successive query expressions with the command-line target specs as the
+            # initial input!
+            if len(query_expr_strings) > 1:
+                raise ValueError("Only one --query argument is currently supported! "
+                                 f"Received: {query_expr_strings}.")
+
+            # TODO: allow returning @union types to avoid this double synchronous engine invocation!
+            exprs = session.product_request(
+                QueryComponentWrapper, [QueryParseInput(s) for s in query_expr_strings]
+            )
+            exprs = [ex.underlying for ex in exprs]
```

Comment:
Alternatively, you might consider not using the engine for parsing? It looks like the parsing is a pure function, so it would likely be cheaper not to bounce in and out like this.

Reply:
+1. Beyond being cheaper perf wise, it would likely be simpler code as pure functions are easier to work with and test than rules.